File: 78-subsets/78-subsets.ts
function subsets(nums: number[]): number[][] {
  const N = nums.length,
        result = [];
  
  backtrack(result, [], 0);
  
  return result;
  
  function backtrack(result: number[][], curr: number[], start: number){
    result.push(curr);
    
    for(let i = start; i < N; i++){
      curr.push(nums[i]);
      backtrack(result, curr.slice(0), i + 1);
      curr.pop();
    }
  }
};

File: 78-subsets/78-subsets.java
class Solution {
	public List<List<Integer>> subsets(int[] nums) {
		List<List<Integer>> ans = new LinkedList<>();

		this.backtrack(0, new Stack<Integer>(), ans, nums);

		return ans;
	}

	private void backtrack(int idx, Stack<Integer> curr, List<List<Integer>> ans, int[] nums) {
		if (idx > nums.length)
			return;

		ans.add(new LinkedList<>(curr));

		for (int i = idx; i < nums.length; i++) {
			curr.push(nums[i]);
			this.backtrack(i + 1, curr, ans, nums);
			curr.pop();
		}

	}
}